src/ocrd/processor/base.py

```diff
@@ -131,50 +175,51 @@ def __init__(
             for res in self.list_all_resources():
                 print(res)
             return
-        if resolve_resource or show_resource:
-            initLogging()
+        if resolve_resource:
             try:
-                res_fname = self.resolve_resource(resolve_resource or show_resource)
+                res = self.resolve_resource(resolve_resource)
+                print(res)
+            except ResourceNotFoundError as e:
+                log = getLogger('ocrd.processor.base')
+                log.critical(e.message)
+                sys.exit(1)
+            return
+        if show_resource:
+            try:
+                self.show_resource(show_resource)
             except ResourceNotFoundError as e:
                 log = getLogger('ocrd.processor.base')
                 log.critical(e.message)
                 sys.exit(1)
-            if resolve_resource:
-                print(res_fname)
-                return
-            fpath = Path(res_fname)
-            if fpath.is_dir():
-                with pushd_popd(fpath):
-                    fileobj = io.BytesIO()
-                    with tarfile.open(fileobj=fileobj, mode='w:gz') as tarball:
-                        tarball.add('.')
-                    fileobj.seek(0)
-                    copyfileobj(fileobj, sys.stdout.buffer)
-            else:
-                sys.stdout.buffer.write(fpath.read_bytes())
             return
         if show_help:
             self.show_help(subcommand=subcommand)
             return
-        self.version = version
         if show_version:
             self.show_version()
             return
         self.workspace = workspace
-        # FIXME HACK would be better to use pushd_popd(self.workspace.directory)
-        # but there is no way to do that in process here since it's an
-        # overridden method. chdir is almost always an anti-pattern.
         if self.workspace:
```

Comment:
Since we did not differentiate between constructor and `setup` until now, `workspace` had to be set in the constructor even for code branches that do not require it (like `--dump-tool-json` or `--show-resource`, hence it needs to be `Optional` typed.

So I have two questions:
a. Are we absolutely certain that `self.workspace` is set before any processing happens
b. Can we convince Python's typing that `self.workspace is not None`?

As for a):
- `self.workspace` is set in the constructor, e.g. by `ocrd_cli_wrap_processor`, so the CLI case is clear
- `run_processor` the processor is instantiated with `workspace=None` but calls `process_workspace` and `self.workspace` is set there.
- `get_cached_processor`: Processor is constructed with `workspace=None` when instance caching and `processor.workspace` is assigned after instantiation.
- `get_processor` without instance caching: `workspace` is passed to the constructor.
- any other case I'm missing?

Can we get rid of `self.workspace` being set in the constructor altogether?

As for b: An easy workaround is to do `assert self.workspace` at the top of the methods that reference it. But it would be even better if we could type `self.workspace` as non-optional, which would be easy if we did not mess with it in the constructor.

Reply:
> a. Are we absolutely certain that `self.workspace` is set before any processing happens

Positive!

> * `run_processor` the processor is instantiated with `workspace=None` but calls `process_workspace` and `self.workspace` is set there.

Yes, and this includes the case where `process` has to be called, because the implementor has not yet adapted to the new API.

> * `get_cached_processor`: Processor is constructed with `workspace=None` when instance caching and `processor.workspace` is assigned after instantiation.
> * `get_processor` without instance caching: `workspace` is passed to the constructor.

Yes, but again, `workspace=None` unless we forgot something.

> * any other case I'm missing?
> Can we get rid of `self.workspace` being set in the constructor altogether?

We do still have processors that are instantiated with a `workspace` argument and then called `process` directly – but only in API tests. Therefore, if we now remove `workspace` from the constructor entirely, we will at least see it fail loudly.

IMHO being bold in changing the constructor (also removing most of the other kwargs and having the decorator call dedicated functions instead) with 3.0 would pay in the long term. The question to me is rather: prior to the first pre-release or later?

Reply:
> b. Can we convince Python's typing that `self.workspace is not None`?

> An easy workaround is to do `assert self.workspace` at the top of the methods that reference it. But it would be even better if we could type `self.workspace` as non-optional, which would be easy if we did not mess with it in the constructor.

We should have made a purely functional API `process_workspace(workspace)` in the first place. By adding `self.workspace`, we felt confident in doing `chdir` in core and accessing `self.workspace` from the implementations. 

With the new API we could start unwinding even the latter. So instead of assertions and a type checked class attribute, we would simply have a type checked (non-optional) argument. But that would break backwards compatibility completely.

Reply:
> Same goes for `self.output_file_grp` BTW. Once we change the constructor more radically, there are plenty of possibilities. See [other thread](https://github.com/OCR-D/core/pull/1240/files#r1713398432).

Oh, no! We still need to have that information in the processors, because for derived images they must construct the output file name in advance – based on the output file grp.

Or maybe we encapsulate those details completely: `Processor.process_workspace` or `Processor.process_page_file` could define two member functions –
1. `Processor.output_file_id(suffix)` – e.g. pass `f"_{segment.id}.IMG-BIN"`; no need for `output_file_id` kwarg in `process_page_pcgts` any more
2. `Processor.output_file_path(filename)` – pass `filename`; no need for `self.output_file_grp` attr any more

Reply:
> Therefore, if we now remove workspace from the constructor entirely, we will at least see it fail loudly.

I am in big favor of that. Workspace has no place in the constructor. Every processor method that involves the need for a workspace, should come as an argument to that method. The workspace itself is something external to the Processor class. It may feel annoying to pass a workspace as an argument everywhere.

Reply:
> IMHO being bold in changing the constructor (also removing most of the other kwargs and having the decorator call dedicated functions instead) with 3.0 would pay in the long term. The question to me is rather: prior to the first pre-release or later?

Agree that it makes sense to be radical at this point. But considering I have a fixed deadline for the pre-release this week, I would not include it in the `3.0.0a1`, but in `3.0.0b1` to be released soon after. If we don't fix this now, I don't see when.

Reply:
Okay. So to sum up:

- Lots of open points
  - [ ] encapsulating `Processor.process_page_pcgts` return type as a class, possibly not even creating image path names in advance
  - [ ] deprecating or even removing the `workspace` arg to the processor constructor
  - [ ] differentiating between processor constructor and all use cases, possibly not even having `input_file_grp` and `output_file_grp` as members at all
  - [ ] doing fileGrp cardinality checks with backing from tool json and in `Processor.process_workspace`
  - [ ] configuring `download_files` via configuration variable
  - [ ] configuring input file failure strategy via configuration variable
  -  [ ] catching exceptions in `Processor.process_page_file`
  - [ ] configuring output file failure strategy via configuration variable
- We have not fully decided which of the agreed / discussed changes we still want in 3.0.0a1. In the extreme, we release the current state, adapt OCR-D/ocrd_kraken#44 and OCR-D/ocrd_tesserocr#216 to take that as a requirement, make these interoperate, and then continue from there.
- We basically had to give up backwards compatibility already (paradoxically because of the way `setup` has been introduced too early). So we'll have to keep doing pre-releases in core for as long as all of our processors do not have either a v3 migration or a v2 freeze. – Unless someone comes up with an idea to prevent duplicate `setup` calls.

Reply:
> Unless someone comes up with an idea to prevent duplicate `setup` calls.

Ah, telling the wood from the trees, of course: If we just avoid defining `Processor.output_file_grp` in the constructor (but only later in `process_workspace`), then we **will** be backwards compatible as long as we want. So a more radical change in core actually helps us relax the interdependencies! (And I'm sure we can do something about ocrd_cis' `Processor.ocrd_tool` usage.)

Reply:
I now separated the processor init and decorator (non-processing calls). I also added the fileGrp cardinality checking to the ocrd-tool.json schema. I will make a PR for spec and update OCR-D/ocrd_kraken#44 and OCR-D/ocrd_tesserocr#216 shortly.

Reply:
Done.

https://github.com/OCR-D/spec/pull/255

Reply:
I'll resolve here and paste the (updated) list of open points in a separate comment below.